kentikapi/cloud_exports_integration_test.go

```diff
@@ -0,0 +1,1028 @@
+			UseBGPDeviceID: "dummy-device-id",
+			DeviceBGPType:  "dummy-device-bgp-type",
+		},
+		CurrentStatus: &models.CloudExportStatus{
+			Status:               "ERROR",
+			ErrorMessage:         "BucketRegionError: incorrect region ...",
+			FlowFound:            pointer.ToBool(false),
+			APIAccess:            pointer.ToBool(true),
+			StorageAccountAccess: pointer.ToBool(false),
+		},
+	}
+}
+
+func newFullAWSCloudExportPayload() *cloudexportpb.CloudExport {
+	ce := newFullCloudExportPayload()
+	ce.Id = awsCloudExportID
+	ce.CloudProvider = "aws"
+	ce.Properties = &cloudexportpb.CloudExport_Aws{
+		Aws: &cloudexportpb.AwsProperties{
+			Bucket:          "dummy-bucket",
+			IamRoleArn:      "arn:aws:iam::003740049406:role/trafficTerraformIngestRole",
+			Region:          "us-east-2",
+			DeleteAfterRead: true,
+			MultipleBuckets: false,
+		},
+	}
+	return ce
+}
+
+func newFullAzureCloudExportPayload() *cloudexportpb.CloudExport {
+	ce := newFullCloudExportPayload()
+	ce.Id = azureCloudExportID
+	ce.CloudProvider = "azure"
+	ce.Properties = &cloudexportpb.CloudExport_Azure{
+		Azure: &cloudexportpb.AzureProperties{
+			Location:                 "dummy-location",
+			ResourceGroup:            "dummy-rg",
+			StorageAccount:           "dummy-sa",
+			SubscriptionId:           "dummy-sid",
+			SecurityPrincipalEnabled: true,
+		},
+	}
+	return ce
+}
+
+func newFullGCECloudExportPayload() *cloudexportpb.CloudExport {
+	ce := newFullCloudExportPayload()
+	ce.Id = gceCloudExportID
+	ce.CloudProvider = "gce"
+	ce.Properties = &cloudexportpb.CloudExport_Gce{
+		Gce: &cloudexportpb.GceProperties{
+			Project:      "dummy-project",
+			Subscription: "dummy-subscription",
+		},
+	}
+	return ce
+}
+
+func newFullIBMCloudExportPayload() *cloudexportpb.CloudExport {
+	ce := newFullCloudExportPayload()
+	ce.Id = ibmCloudExportID
+	ce.CloudProvider = "ibm"
+	ce.Properties = &cloudexportpb.CloudExport_Ibm{
+		Ibm: &cloudexportpb.IbmProperties{
+			Bucket: "dummy-bucket",
+		},
+	}
+	return ce
+}
+
+func newFullCloudExportPayload() *cloudexportpb.CloudExport {
+	return &cloudexportpb.CloudExport{
+		Type:        cloudexportpb.CloudExportType_CLOUD_EXPORT_TYPE_KENTIK_MANAGED,
+		Enabled:     true,
+		Name:        "full-export",
+		Description: "Export with all fields set", // including read-only fields
+		ApiRoot:     "https://api.kentik.com",
+		FlowDest:    "https://flow.kentik.com",
+		PlanId:      "11467",
+		Bgp: &cloudexportpb.BgpProperties{
+			ApplyBgp:       true,
+			UseBgpDeviceId: "dummy-device-id",
+			DeviceBgpType:  "dummy-device-bgp-type",
+		},
+		CurrentStatus: &cloudexportpb.Status{
+			Status:               "ERROR",
+			ErrorMessage:         "BucketRegionError: incorrect region ...",
+			FlowFound:            &wrapperspb.BoolValue{Value: false},
+			ApiAccess:            &wrapperspb.BoolValue{Value: true},
+			StorageAccountAccess: &wrapperspb.BoolValue{Value: false},
+		},
+	}
+}
+
+func codePtr(c codes.Code) *codes.Code {
+	return &c
+}
```

Comment:
Given that both CloudExport grpc client and server are auto-generated from the same grpc protocol files, I'm not quite sure if such integration testing presents much profit; Im afraid the manually written integration test is more prone to errors than the auto-generated client-server code under test. 
The part that might deserve testing is data transformation between models <-> pb and this can be done as unit tests. Although not that much necessary; we are simply repacking data from one GO structure to another, not dealing with any savage untyped json data. What do you think?

Reply:
I have to end for today, but in short: this integration test tests also conversions from models to pb and back. I could test that also via unit test. It wouldn't be easier for me and it wouldn't be noticably more performant.

If I have to choose between integration test and unit test, I pick integration test, unless it is noticably slower/harder to write. Because it tests more interactions by definition

Reply:
OK, it's already in place anyway, I just think it's an overkill

Reply:
It might be an overkill. I frankly don't know if UTs would be easier to write for me.